add bundlers document (#11472)
microsoft/FluidFramework: docs/content/docs/bundlers.md

---
title: Fluid and Bundlers
aliases:
  - "/start/bundlers/"
---

> The following are known bundler issues, answers to some of the most commonly asked questions about the Fluid
> Framework and Bundlers.

Reviewer: ```suggestion
This article describes some known issues with bundlers and answers some commonly asked questions about the Fluid
Framework and bundlers.
```

## Known Errors with Webpack 5

> We know the following packages cause errors; assert, buffer, events and url.

Reviewer: It's not clear what these 4 things are? Are they part of FF? Are they part of WebPack? Are they popular JS libraries?

Author: What about calling them core NodeJS Modules?

> We know the following packages cause errors; assert, buffer, events and url.

Reviewer 2: It would probably be worth adding a brief introductory statement noting that these may be encountered when migrating from Webpack 4 to Webpack 5. The target audience of this doc is specifically users who are already using us with Webpack v4 and are upgrading.

Reviewer 3: > The target audience of this doc is specifically users who are already using us with Webpack v4 and are upgrading.

Not necessarily. The people I am most concerned about are those that are adding Fluid to an existing project or using a scaffolding tool like create-react-app. People that are upgrading Webpack 4 -> 5 know about polyfills, usually. People starting fresh with Webpack 5 often expect it to "just work."

Author: Is there anything else to do here or should I mark this as resolved?

Reviewer 2: I think we're good then. I misunderstood the situation 🙂

> We know the following packages cause errors; assert, buffer, events and url.

Reviewer 3: ```suggestion
The Fluid Framework uses some node.js core modules that do not exist in the browser. webpack 4 and below used to include polyfills for node.js core modules by default. This is no longer the case in webpack 5 and beyond, so these modules must be added to your project and webpack 5 config.
```

> This issue causes compiling any project that uses these libraries to fail. Upgrading to Webpack 5 removed the automatic NodeJS polyfills. We now have to manually install the missing packages and add the fallback.

Reviewer: JS is not a compiled language, so what does "compiling" mean? "bundling"? "transpiling from TypeScript"?

Reviewer 3: "bundling" is the better term here.

> To fix this, run the command `npm install -D <package>` and add the fallback `resolve.fallback: { "<package>": require.resolve("<package>/") }` to the affected package's webpack.config.js. An example of this is located [here](https://github.com/microsoft/FluidFramework/blob/a4c38234a920abe9b54b1c26a14c0a8e430cd3fa/packages/tools/webpack-fluid-loader/webpack.config.js#L37)

Reviewer: ```suggestion
To fix this take these steps: 

1. Run the command `npm install -D <package>`
2. In the webpack.config.js file, add the following `fallback` property to the `resolve` object: 

    ```javascript
    fallback: { 
            "<package>": require.resolve("<package>/") 
        }
    ``` 

    Add an additional `"<package>"` property to the `fallback` object for each of the four problematic packages that you are using. An example of this is located [here](https://github.com/microsoft/FluidFramework/blob/a4c38234a920abe9b54b1c26a14c0a8e430cd3fa/packages/tools/webpack-fluid-loader/webpack.config.js#L37)
```

Error:
```bash

> BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.

Reviewer: Again here, is this an internal doc or one aimed at external audiences? If it is external, why call this a breaking change? Why assume they already had an FF project with WP 4?

Reviewer 2: Is this example error output we get from Webpack? If so, we should clarify that a bit more. We definitely want to be clear that we are surfacing common issues from *Webpack's* breaking change, and not noting a breaking change of our own.

Reviewer 3: Yes, this is what webpack spits out when you have dependencies on node core modules. I think we need to have the error _somewhere_ in the docs to help people find it.

Author: Is there anything else to do here or should I mark this as resolved?

Reviewer 2: I think we at least need to clarify the label we're using before the sample error. Right now it just says "Error:", but I think we need give a bit more context into when the error is expected to occur.

Reviewer 2: Change looks good to me. @Rick-Kirkham do you have any remaining concerns here?
